### PR TITLE
feat: rebuild AliveStatements in terms of LLVM.IntW

### PR DESCRIPTION
**Pull request by tobiasgrosser**
feat: rebuild AliveStatements in terms of LLVM.IntW

Add a new script to derive AliveStatements.lean from AliveAutoGenerated.lean. In the process, we move the content of AliveStatements to work on LLVM.IntW, instead of bitvectors. This has the advantage that AliveAutoGenerated.lean only eliminates the SSA overhead, which should be a deterministic process. Heuristic-based solutions to eliminate the option monad for modeling undefined behavior or to prove the theorems can then be limited to AliveStatements.lean.

This fixes a couple of broken proofs in AliveAutoGenerated.lean and leaves us with only two broken proofs that will be fixed by either https://github.com/opencompl/ssa/pull/225 or https://github.com/opencompl/ssa/pull/247.

 Also:
 - Remove old extract_alive_statements.sh file
 - CI support to verify that AliveStatements.lean is up-to-date.

### CODE REVIEW

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    if len(error) >= 7:
+        stmt += "\n" + error[4]
+    if len(error) >= 8:
+        stmt += "\n" + error[5]
+    if len(error) >= 9:
+        stmt += "\n" + error[6]
+    if len(error) >= 10:
+        stmt += "\n" + error[7]
+    if len(error) >= 11:
+        stmt += "\n" + error[8]
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:58]: This error checking code based on length looks totally crazy at first glance. I assume the length indicates something, can I have a descriptive constant name, and better yet, an explanation of what's going on here?

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import SSA.Projects.InstCombine.ForStd
+import SSA.Projects.InstCombine.ForMathlib
+import SSA.Projects.InstCombine.LLVM.Semantics
+import Std.Data.BitVec
+import Mathlib.Data.BitVec.Lemmas
+
+open LLVM
+open BitVec
+"""
+writeOutput(preamble, statements, "AliveStatements.lean")
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:98]: Add a `if __name__ == "__main_"` to allow for debugging by importing into a REPL. 

```suggestion
if __name__ == "__main__":
  f = open("AliveAutoGenerated.lean", "r")
  lines = f.readlines()
  preamble, proofs = getProofs(lines)
  statements = filterProofs(preamble, proofs)
  # writeOutput(preamble, working_proofs, "AliveAutoGenerated.lean")
  preamble = """
  /-
  Released under Apache 2.0 license as described in the file LICENSE.
  -/
  import SSA.Projects.InstCombine.TacticAuto
  import SSA.Projects.InstCombine.ForStd
  import SSA.Projects.InstCombine.ForMathlib
  import SSA.Projects.InstCombine.LLVM.Semantics
  import Std.Data.BitVec
  import Mathlib.Data.BitVec.Lemmas
  
  open LLVM
  open BitVec
  """
  writeOutput(preamble, statements, "AliveStatements.lean")
```

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        if line.startswith("  apply"):
+            rewritten.append("  print_goal_as_error")
+            name = line.split(" ")[3][0:-1]
+        else:
+            rewritten.append(line)
+    f.write("".join(rewritten))
+
+    f.close()
+    
+    import subprocess
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:35]: Preferably, keep imports toplevel.
```suggestion
```

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+# Released under Apache 2.0 license as described in the file LICENSE.
+
+import re
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:4]:
```suggestion
import re
import subprocess
from typing import *
```

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+def getProofs(lines):
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:6]:
```suggestion
def getProofs(lines : List[str]) -> Tuple[List[str],  List[List[str]]]:
  """
  Breaks the file apart into subarrays.
  Each subarray starts with the line `-- Name`.
  This effectively returns a list of sequences of lines, where each sequence of lines corresponds to a proof.
  Peels the first subarray (the "preamble") out, which contains the imports and other setup before the proofs.
  """
```

File: SSA/Projects/InstCombine/update_alive_statements.py (hunk 4 above)
Comment (bollu) [reply SSA/Projects/InstCombine/update_alive_statements.py:4]: I'm not as strict as sasha, but I would like some type information please!

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    for line in lines:
+        if line.startswith("-- Name"):
+            proofs.append(proof)
+            proof = []
+        proof.append(line)
+    proofs.append(proof)
+
+    return proofs[0], proofs[1:]
+
+def getStatement(preamble, proof):
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:19]:
```suggestion
def getStatement(preamble : List[str], proof : List[str]) -> List[str]:
  """
  Produces the proof state produced by running the `proof`. 
  Uses the preamble to create a valid Lean file.
  """
```

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    stmt += " := by\n  simp_alive_undef\n  simp_alive_ops\n  simp_alive_case_bash\n  try alive_auto\n  try sorry"
+    print(stmt)
+    return stmt
+
+def filterProofs(preamble, proofs):
+    statements = []
+    for proof in proofs:
+        print("processing: " + proof[0])
+        stmt = getStatement(preamble, proof)
+        statements.append(stmt)
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:69]: I'm a little confused, don't we need a check that says: 
```suggestion
        stmt = getStatement(preamble, proof)
        if not stmt: continue
        statements.append(stmt)
```

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        f.write("".join(preamble))
+        for proof in proofs:
+            f.write("".join(proof))
+
+f = open("AliveAutoGenerated.lean", "r")
+lines = f.readlines()
+preamble, proofs = getProofs(lines)
+statements = filterProofs(preamble, proofs)
+# writeOutput(preamble, working_proofs, "AliveAutoGenerated.lean")
+preamble = """
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:84]: Why do we both grab the preamble above in `getProofs`, as well as have a handwritten preamble?

**Review by bollu — COMMENTED** [review]
Thanks a lot for working on this @tobiasgrosser!

Could you also edit the CI to run this script?

File: SSA/Projects/InstCombine/update_alive_statements.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    if len(error) >= 10:
+        stmt += "\n" + error[7]
+    if len(error) >= 11:
+        stmt += "\n" + error[8]
+    print("\n")
+    stmt += " := by\n  simp_alive_undef\n  simp_alive_ops\n  simp_alive_case_bash\n  try alive_auto\n  try sorry"
+    print(stmt)
+    return stmt
+
+def filterProofs(preamble, proofs):
```
Comment (bollu) [inline SSA/Projects/InstCombine/update_alive_statements.py:64]:
```suggestion
def filterProofs(preamble : List[str], proofs : List[List[str]]) -> List[str]:
  """
  Returns a list of statements, corresponding to each of the proofs.  
  """
```

File: SSA/Projects/InstCombine/update_alive_statements.py (hunk 1 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/update_alive_statements.py:69]: Fixed.

File: SSA/Projects/InstCombine/update_alive_statements.py (hunk 8 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/update_alive_statements.py]: Fixed.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -48,6 +48,11 @@ jobs:
           sudo apt install llvm-15 # for opt-15, used to simplify LLVM for reference semantics.
           cd test/bruteforce-correctness && ./run.sh
 
+      - name: 'Check for changes in AliveStatements.lean'
+        run: |
+          (cd SSA/Projects/InstCombine/; update_alive_statements.py)
+	  git diff | grep .
```
Comment (bollu) [inline .github/workflows/ci.yml:54]: @tobiasgrosser Might I suggest 

```
set -e expandtab
set tabstop=4
```

in your vimrc?

```suggestion
          (cd SSA/Projects/InstCombine/; update_alive_statements.py)
          git diff | grep .
```

Comment (tobiasgrosser) [reply .github/workflows/ci.yml]: Done.

**Comment by bollu** [comment]
Lgtm, I’ve reviewed the patch and it makes sense. I’d appreciate a line or two of comments in the cryptic workflow incantation of git diff.

**Comment by tobiasgrosser** [comment]
> Lgtm, I’ve reviewed the patch and it makes sense. I’d appreciate a line or two of comments in the cryptic workflow incantation of git diff.

Done.

**Pull request merged**